File: java/money/ch14/src/main/java/com/thesoftwaregorilla/tdd/money/Bank.java
package com.thesoftwaregorilla.tdd.money;


import java.util.HashMap;

public class Bank {

    private final HashMap<CurrencyPair, Integer> rates = new HashMap<>();

    public Money reduce(Expression source, String to) {
        return source.reduce(this, to);
    }

    public int rate(String from, String to) {
        if (from.equals(to)) {
            return 1;
        }
        Integer rate = rates.get(new CurrencyPair(from, to));
        return rate != null ? rate : 0;
    }

    public void addRate(String from, String to, int rate) {
        rates.put(new CurrencyPair(from, to), rate);
    }

}

File: cs/ch15/Money/Bank.cs
namespace TheSoftwareGorilla.TDD.Money;

public class Bank
{
    private Dictionary<CurrencyPair, int> _rates = new Dictionary<CurrencyPair, int>();
    public Money Reduce(Expression source, string to)
    {
        return source.Reduce(this, to);
    }

    public int Rate(string from, string to)
    {
        if (from.Equals(to))
        {
            return 1;
        }
        int rate = _rates.TryGetValue(new CurrencyPair(from, to), out int value) ? value : 0;
        return rate;
    }

    public void AddRate(string from, string to, int rate)
    {
        _rates.Add(new CurrencyPair(from, to), rate);
    }
}